geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geohash/GeohashUtils.scala

```diff
@@ -359,6 +359,73 @@ object GeohashUtils
   }
 
   /**
+   * Find GeoHash instance with maximum precision that covers envelope defined by two points.
+   *
+   * @param ll Low left point of bounding box
+   * @param ur Up right point of bounding box
+   * @return GeoHash instance
+   */
+  def getMBGH(ll: Point, ur: Point): GeoHash = {
+    val width = ur.getX - ll.getX
+    val height = ur.getY - ll.getY
+    require(width >= 0 && height >= 0, s"Wrong width $width and height $height of input bounding box, cannot process")
+
+    (60 to 0 by -5).foreach(prec => {
+      val lonDelta = GeoHash.longitudeDeltaForPrecision(prec)
+      val latDelta = GeoHash.latitudeDeltaForPrecision(prec)
+      if (lonDelta >= width && latDelta >= height) {
+        val geo = GeoHash(ll.getX, ll.getY, prec)
+        if (geo.bbox.covers(ur)) return geo
+      }
+    })
+    null
+  }
+
+  def getMBGH(env: Envelope): GeoHash =
+    GeohashUtils.getMBGH(env.getMinX, env.getMaxX, env.getMinY, env.getMaxY)
+
+  def getMBGH(bbox: BoundingBox): GeoHash =
+    getMBGH(bbox.getMinX, bbox.getMaxX, bbox.getMinY, bbox.getMaxY)
+
+  def getMBGH(minX: Double, maxX: Double, minY: Double, maxY: Double): GeoHash =
+    getMBGH(GeoHash.factory.createPoint(new Coordinate(minX, minY)),
+      GeoHash.factory.createPoint(new Coordinate(maxX, maxY)))
+
+  // TODO: WCS: Make the ClosestAcceptableGeoHash algorithm more robust -- GEOMESA-590
+  // TODO: WCS: Using this in the rowID will require the query bounding box to be expanded
+  def getClosestAcceptableGeoHash(env: Envelope): Option[GeoHash] = {
+    getClosestAcceptableGeoHash(BoundingBox(env))
+  }
+
+  /**
+   * NOTE: None is the whole world
+   * @param bbox the BoundingBox to which the closest acceptable GeoHash will be derived from.
+   * @return
+   */
+  def getClosestAcceptableGeoHash(bbox: BoundingBox): Option[GeoHash] = {
+    val prec = calculatePrecision(bbox)
+    if (prec >= 0 ) {
+      // The GeoHash precision must be some multiple of 5 to be correctly represented via the 32-bit GeoHash encoding
+      val gh = getClosestAcceptableGeoHash(bbox, prec)
+      prec % 5 match {
+        case 0               => Some(gh)
+        case _ if (prec > 5) => Some(GeoHash(gh.hash.dropRight(1)))
```

Comment:
the default case is always needed for case matches. plus this case is hit when the precision in less than 5.